do/csi/driver/controller.go

```diff
@@ -3,48 +3,296 @@ package driver
+	for {
+		vols, resp, err := d.doClient.Storage.ListVolumes(ctx, listOpts)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, vol := range vols {
+			volumes = append(volumes, vol)
+		}
+
+		if resp.Links == nil || resp.Links.IsLastPage() {
+			page, err := resp.Links.CurrentPage()
+			if err != nil {
+				return nil, err
+			}
+			// save this for the response
+			lastPage = page
+			break
+		}
+
+		page, err := resp.Links.CurrentPage()
+		if err != nil {
+			return nil, err
+		}
+
+		listOpts.ListOptions.Page = page + 1
+	}
+
+	var entries []*csi.ListVolumesResponse_Entry
+	for _, vol := range volumes {
+		entries = append(entries, &csi.ListVolumesResponse_Entry{
+			Volume: &csi.Volume{
+				Id:            vol.ID,
+				CapacityBytes: vol.SizeGigaBytes * GB,
+			},
+		})
+	}
+
+	// TODO(arslan): check that the NextToken logic works fine, might be racy
+	return &csi.ListVolumesResponse{
+		Entries:   entries,
+		NextToken: strconv.Itoa(lastPage),
+	}, nil
 }
 
-// GetCapacity ...
-func (d *Driver) GetCapacity(context.Context, *csi.GetCapacityRequest) (*csi.GetCapacityResponse, error) {
-	return nil, errors.New("not implemented")
+// GetCapacity returns the capacity of the storage pool
+func (d *Driver) GetCapacity(ctx context.Context, req *csi.GetCapacityRequest) (*csi.GetCapacityResponse, error) {
+	// TODO(arslan): check if we can provide this information somehow
+	return nil, status.Error(codes.Unimplemented, "")
 }
 
-// ControllerGetCapabilities ...
-func (d *Driver) ControllerGetCapabilities(context.Context, *csi.ControllerGetCapabilitiesRequest) (*csi.ControllerGetCapabilitiesResponse, error) {
-	return nil, errors.New("not implemented")
+// ControllerGetCapabilities returns the capabilities of the controller service.
+func (d *Driver) ControllerGetCapabilities(ctx context.Context, req *csi.ControllerGetCapabilitiesRequest) (*csi.ControllerGetCapabilitiesResponse, error) {
+	newCap := func(cap csi.ControllerServiceCapability_RPC_Type) *csi.ControllerServiceCapability {
+		return &csi.ControllerServiceCapability{
+			Type: &csi.ControllerServiceCapability_Rpc{
+				Rpc: &csi.ControllerServiceCapability_RPC{
+					Type: cap,
+				},
+			},
+		}
+	}
+
+	// TODO(arslan): checkout if the capabilities are worth supporting
+	var caps []*csi.ControllerServiceCapability
+	for _, cap := range []csi.ControllerServiceCapability_RPC_Type{
+		csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME,
+		csi.ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME,
+		csi.ControllerServiceCapability_RPC_LIST_VOLUMES,
+	} {
+		caps = append(caps, newCap(cap))
+	}
+
+	return &csi.ControllerGetCapabilitiesResponse{
+		Capabilities: caps,
+	}, nil
+}
+
+// extractStorage extracts the storage size in GB from the given capacity
+// range. If the capacity range is not satisfied it returns the default volume
+// size.
+func extractStorage(capRange *csi.CapacityRange) (int64, error) {
+	if capRange == nil {
+		return defaultVolumeSizeInGB, nil
```

Comment:
This happens if the CO (in our case Kubernetes) doesn't set the size of the requested volume. In this case we need to define a standard our self. I've set it to 16GB, but this might change to anything else.